docs/user_guide/ml.md

```diff
@@ -1,8 +1,254 @@
+Remember that you already have a copy of all the input and target features in your parquet files used for training.
+In General it is not advised to create a new separate copy of these, otherwise it can be convenient to have everything within one place.
+Thus, the choice to include these columns is your choice.
+
+## uses:
+In `uses` you define the columns that are needed by your machine learning model, and there that are not filtered out.
+In this case we want to preserve the input and target features, as well as some weights:
+```{literalinclude} ./ml_code.py
+:language: python
+:pyobject: MLModel.uses
+```
+
+## output:
+In the {py:meth}`~columnflow.ml.MLModel.output` function, you define your local target directory that your current model instance will have access to.
+
+Since machine learning in columnflow uses k-fold by default, it is a good idea to have a separate directory for each fold, and this should be reflected in the {py:meth}`~columnflow.ml.MLModel.output` path.
+It is of good practice to use `task.target` to be that your "machine-learning-task" related files are saved in the same directory.
+In this example we want to save each fold separately e.g:
+```{literalinclude} ./ml_code.py
+:language: python
+:start-at: def output
+:end-at: return target
+```
+
+
+## open_model:
+In the {py:meth}`~columnflow.ml.MLModel.open_model` function, you implement the loading of the trained model and if necessary its configuration, so it is ready to use for {py:meth}`~columnflow.ml.MLModel.evaluate`.
+This does not define how the model is build for {py:meth}`~columnflow.ml.MLModel.train`.
+
+The `target` parameter represents the local path to the models directory.
+In the the following example a TensorFlow model saved with Keras API is loaded and returned, and no further configuration happens:
+```{literalinclude} ./ml_code.py
+:language: python
+:start-at: def open_model
+:end-at: return loaded_model
+```
+
+## train:
+In the {py:meth}`~columnflow.ml.MLModel.train` function, you implement the initialization of your models and the training loop.
+The `task` corresponding to the models training is {py:class}`~columnflow.tasks.ml.MLTraining`.
+By default {py:meth}`~columnflow.ml.MLModel.train` has access to the location of the models `inputs` and `outputs`.
+
+In columnflow, k-fold is enabled by default.
+The `Self` argument in {py:meth}`~columnflow.ml.MLModel.train` referes to the instance of the fold.
+Using `Self`, you have also access to the entire `analysis_inst`ance the `config_inst`ance of the current fold, and to all the derived parameters of your model.
+
+With this information, you can call and prepare the columns to be used by the model for training.
+In the following exmaple a very simple dummy trainings loop is performed using Keras fit function.
+```{literalinclude} ./ml_code.py
+:language: python
+:start-at: def train
+:end-at: return
+```
```

Comment:
Side note: you should also explain `self.build_model()`, `self.open_input_files()` and `self.prepare_events()` somewhere. API is also fine of course

Reply:
These are helper functions I suggest people should design, thus they are not part of CF. This was the reason why I have no further explanation of these functions.